azure/batch/scripts/predict_on_catalog.py

```diff
@@ -50,46 +50,46 @@ def open_image_as_rgb(img):
 class PredictionGalaxyDataset(galaxy_dataset.GalaxyDataset):
   # override the default class implementation for predictions that download from URL
   def __init__(self, catalog: pd.DataFrame, label_cols=None, transform=None, target_transform=None):
-      self.catalog = catalog
-
       self.label_cols = label_cols
       self.transform = transform
       self.target_transform = target_transform
 
-  def __getitem__(self, idx):
-      galaxy = self.catalog.iloc[idx]
-      # load the data from the remote image URL
-      url = galaxy['image_url']
-      try:
-          # streaming the file as it is used (saves on memory)
-          logging.debug('Downloading url: {}'.format(url))
-          # use retries on requests if we have flaky networks
-          response = requests_retry_session().get(url, stream=True)
-          # ensure we raise other response errors like 404 and 500 etc
-          # Note: we don't retry on errors that aren't in the `status_forcelist`, instead we fast fail!
-          response.raise_for_status()
-          image = open_image_as_rgb(response.raw)
-      except Exception as e:
-          # add some logging on the failed url
-          logging.critical('Cannot load {}'.format(url))
-          # and make sure we reraise the error
-          raise e
-
-      # avoid the label lookups as they aren't used in the prediction
-      label = []
-
-      # logging.info((image.shape, torch.max(image), image.dtype, label))  # always 0-255 uint8
-
-      if self.transform:
-          # a CHW tensor, which torchvision wants. May change to PIL image.
-          image = self.transform(image)
-
-      if self.target_transform:
-          label = self.target_transform(label)
-
-      # logging.info((image.shape, torch.max(image), image.dtype, label))  #  should be 0-1 float
-      return image
+      self.data = []
+      valid_rows = []
+      for _, row in catalog.iterrows():
```

Comment:
Your intention is to have this loop over `catalog` rows, which fills the `data` list, at time that `PredictionGalaxyDataset` is instantiated (via `__init__()`)? I suspect that's going to suffer from the same problem the url_ok() solution did, where the instantiation runs multiple times (thereby looping over the full dataset), not just when the predict code requests make a `__getitem__()` call.